### PR TITLE
*: update dependencies

### PR DESCRIPTION
**Pull request by BusyJay**
## What have you changed? (mandatory)

This PR changes the dependencies of kvproto, raft-rs and tipb to master
branch. And it also upgrade grpcio to the latest alpha version.

## What are the type of the changes? (mandatory)

- Engineering

## How has this PR been tested? (mandatory)

unit tests

## Does this PR affect documentation (docs) or release note? (mandatory)

No.

## Does this PR affect tidb-ansible update? (mandatory)

No.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -213,8 +213,8 @@ clippy: pre-clippy
 		-A clippy::implicit_hasher -A clippy::large_enum_variant -A clippy::new_without_default \
 		-A clippy::neg_cmp_op_on_partial_ord -A clippy::too_many_arguments \
 		-A clippy::excessive_precision -A clippy::collapsible_if -A clippy::blacklisted_name \
-		-A clippy::needless_range_loop -D rust-2018-idioms -A clippy::redundant_closure \
```
Comment (BusyJay) [inline Makefile]: Denying rust-2018-idioms can lead to compilation of kvproto, and it can't be fixed by explicitly allow on kvproto's source. Do you have any ideas? /cc @nrc

**Comment by BusyJay** [comment]
I suggest not to merge it until pingcap/grpc-rs#349, which can solve the pain of a common building error.

File: Cargo.toml [hunk 2]
```diff
@@ -103,8 +103,8 @@ zipf = "5.0.1"
 bitflags = "1.0.1"
 fail = "0.3"
 uuid = { version = "0.6", features = [ "serde", "v4" ] }
-grpcio = { version = "0.4", features = [ "secure" ] }
-raft = "0.4.3"
+grpcio = { version = "0.5.0-alpha.2", features = [ "secure" ] }
+raft = { git = "https://github.com/pingcap/raft-rs.git" }
```
Comment (nrc) [inline Cargo.toml]: Can we specify a `rev` for these git deps? I think it is too easy to break otherwise.

File: Makefile (hunk 1 above)
Comment (nrc) [reply Makefile]: I think adding `#[allow(bare_trait_objects)]` to `lib.rs` in kvproto should fix the errors. There is a bug in Clippy where that is ignored for path deps, but with git deps it should be ok (https://github.com/rust-lang/rust-clippy/issues/1066). If that doesn't work, then I think just removing the deny here is fine - it's not a correctness issue and the idioms are generally easy to fix.

File: src/import/service.rs [hunk 3]
```diff
@@ -1,12 +1,12 @@
 // Copyright 2018 TiKV Project Authors. Licensed under Apache-2.0.
 
 use futures::Future;
-use grpcio::{RpcContext, RpcStatus, RpcStatusCode, UnarySink};
+use grpcio::{RpcContext, RpcStatus, RpcStatusCode::*, UnarySink};
```
Comment (nrc) [inline src/import/service.rs]: Could we avoid the glob import of enum variants? I think it is a bad practice because it makes code less readable (except for very well known enums like `Option`)

File: src/raftstore/store/fsm/apply.rs [hunk 4]
```diff
@@ -2637,7 +2640,7 @@ impl ApplyFsm {
             .iter()
             .any(|res| res.region_id == self.delegate.region_id())
             && self.delegate.last_sync_apply_index != applied_index;
-        fail_point!("apply_on_handle_snapshot_sync", |_| { need_sync = true });
+        (|| fail_point!("apply_on_handle_snapshot_sync", |_| { need_sync = true }))();
```
Comment (nrc) [inline src/raftstore/store/fsm/apply.rs]: Why is this change needed?

File: src/raftstore/store/peer_storage.rs [hunk 5]
```diff
@@ -453,8 +453,13 @@ impl Storage for PeerStorage {
         self.initial_state()
     }
 
-    fn entries(&self, low: u64, high: u64, max_size: u64) -> raft::Result<Vec<Entry>> {
-        self.entries(low, high, max_size)
+    fn entries(
+        &self,
+        low: u64,
+        high: u64,
+        max_size: impl Into<Option<u64>>,
```
Comment (nrc) [inline src/raftstore/store/peer_storage.rs]: IMO the `impl Into` is not worth it, just use an `Option` and the callers can write `Some` if they need too. You can keep it if you really like it.

File: src/server/raft_client.rs [hunk 6]
```diff
@@ -13,7 +13,7 @@ use crate::server::transport::RaftStoreRouter;
 use crossbeam::channel::SendError;
 use futures::{future, stream, Future, Poll, Sink, Stream};
 use grpcio::{
-    ChannelBuilder, Environment, Error as GrpcError, RpcStatus, RpcStatusCode, WriteFlags,
+    ChannelBuilder, Environment, Error as GrpcError, RpcStatus, RpcStatusCode::*, WriteFlags,
```
Comment (nrc) [inline src/server/raft_client.rs]: Same as above

File: src/server/service/debug.rs [hunk 7]
```diff
@@ -7,7 +7,7 @@ use futures::sync::oneshot;
 use futures::{future, stream, Future, Stream};
 use futures_cpupool::{Builder, CpuPool};
 use grpcio::{Error as GrpcError, WriteFlags};
-use grpcio::{RpcContext, RpcStatus, RpcStatusCode, ServerStreamingSink, UnarySink};
+use grpcio::{RpcContext, RpcStatus, RpcStatusCode::*, ServerStreamingSink, UnarySink};
```
Comment (nrc) [inline src/server/service/debug.rs]: and again

File: src/server/service/kv.rs [hunk 8]
```diff
@@ -18,7 +18,7 @@ use futures::executor::{self, Notify, Spawn};
 use futures::{future, Async, Future, Sink, Stream};
 use grpcio::{
     ClientStreamingSink, DuplexSink, Error as GrpcError, RequestStream, RpcContext, RpcStatus,
-    RpcStatusCode, ServerStreamingSink, UnarySink, WriteFlags,
+    RpcStatusCode::*, ServerStreamingSink, UnarySink, WriteFlags,
```
Comment (nrc) [inline src/server/service/kv.rs]: and again

File: src/server/service/kv.rs [hunk 9]
```diff
@@ -72,13 +72,7 @@ impl<T: RaftStoreRouter + 'static, E: Engine> Service<T, E> {
...
     }
 
-    fn send_fail_status<M>(
-        &self,
-        ctx: RpcContext<'_>,
-        sink: UnarySink<M>,
-        err: Error,
-        code: RpcStatusCode,
-    ) {
+    fn send_fail_status<M>(&self, ctx: RpcContext<'_>, sink: UnarySink<M>, err: Error, code: i32) {
```
Comment (nrc) [inline src/server/service/kv.rs]: Not sure if we can do anything about it, but changing from a named type to i32 is a readability regression. If we have to do this, could you give the variable a more descriptive name or add a comment to describe what kind of a code it is and what the valid values are.

File: src/server/snap.rs [hunk 10]
```diff
@@ -8,7 +8,7 @@ use std::time::{Duration, Instant};
 use futures::{future, Async, Future, Poll, Stream};
 use futures_cpupool::{Builder as CpuPoolBuilder, CpuPool};
 use grpcio::{
-    ChannelBuilder, ClientStreamingSink, Environment, RequestStream, RpcStatus, RpcStatusCode,
+    ChannelBuilder, ClientStreamingSink, Environment, RequestStream, RpcStatus, RpcStatusCode::*,
```
Comment (nrc) [inline src/server/snap.rs]: and again

File: src/storage/lock_manager/deadlock.rs [hunk 11]
```diff
@@ -13,7 +13,7 @@ use crate::tikv_util::security::SecurityManager;
 use crate::tikv_util::worker::{FutureRunnable, FutureScheduler, Stopped};
 use futures::{Future, Sink, Stream};
 use grpcio::{
-    self, DuplexSink, RequestStream, RpcContext, RpcStatus, RpcStatusCode, UnarySink, WriteFlags,
+    self, DuplexSink, RequestStream, RpcContext, RpcStatus, RpcStatusCode::*, UnarySink, WriteFlags,
```
Comment (nrc) [inline src/storage/lock_manager/deadlock.rs]: again

File: tests/integrations/pd/mock/server.rs [hunk 12]
```diff
@@ -6,7 +6,7 @@ use std::time::Duration;
 
 use futures::{Future, Sink, Stream};
 use grpcio::{
-    DuplexSink, EnvBuilder, RequestStream, RpcContext, RpcStatus, RpcStatusCode,
+    DuplexSink, EnvBuilder, RequestStream, RpcContext, RpcStatus, RpcStatusCode::*,
```
Comment (nrc) [inline tests/integrations/pd/mock/server.rs]: again

File: tests/integrations/raftstore/test_service.rs [hunk 13]
```diff
@@ -4,7 +4,7 @@ use std::path::Path;
 use std::sync::*;
 
 use futures::{future, Future, Stream};
-use grpcio::{ChannelBuilder, Environment, Error, RpcStatusCode};
+use grpcio::{ChannelBuilder, Environment, Error, RpcStatusCode::*};
```
Comment (nrc) [inline tests/integrations/raftstore/test_service.rs]: ^

File: tests/integrations/server/mod.rs [hunk 14]
```diff
@@ -6,6 +6,7 @@ mod raft_client;
 use std::sync::Arc;
 
 use futures::Future;
+use grpcio::RpcStatusCode::*;
```
Comment (nrc) [inline tests/integrations/server/mod.rs]: ^

File: tests/integrations/server/raft_client.rs [hunk 15]
```diff
@@ -6,7 +6,8 @@ use std::{thread, time};
 
 use futures::{Future, Stream};
 use grpcio::{
-    ClientStreamingSink, Environment, RequestStream, RpcContext, RpcStatus, RpcStatusCode, Server,
+    ClientStreamingSink, Environment, RequestStream, RpcContext, RpcStatus, RpcStatusCode::*,
```
Comment (nrc) [inline tests/integrations/server/raft_client.rs]: ^

File: Cargo.toml (hunk 2 above)
Comment (BusyJay) [reply Cargo.toml]: No, it won't. Version is locked in Cargo.lock. Besides, given a version requires kvproto also give a version, so upgrade raft means two repos both have to be updated. But I do think we should depend on a version once it's released, maybe 0.5.0-alpha.2.

File: Makefile (hunk 1 above)
Comment (BusyJay) [reply Makefile]: > I think adding #[allow(bare_trait_objects)] to lib.rs

That was the first thing I tried, and it didn't work.

File: src/import/service.rs (hunk 3 above)
Comment (BusyJay) [reply src/import/service.rs]: It's a module rather than enum. It's generated by rust-bindgens.

File: src/raftstore/store/fsm/apply.rs (hunk 4 above)
Comment (BusyJay) [reply src/raftstore/store/fsm/apply.rs]: It's just the original code which was modified wrongly by #4903. Actually the tests should always fail without this revert, however there is a bug in our CI, which skips the fail cases.

File: src/raftstore/store/peer_storage.rs (hunk 5 above)
Comment (BusyJay) [reply src/raftstore/store/peer_storage.rs]: I think write pure number is quite handy. Besides, it's a method of the trait Storage, I don't think I can change the definition in TiKV.

File: src/server/service/kv.rs (hunk 9 above)
Comment (BusyJay) [reply src/server/service/kv.rs]: I prefer to leave it to grpcio crates.

Comment (BusyJay) [reply src/server/service/kv.rs]: @hunterlxt Do you want to give it a try? You can refer to the http crate for how to define a status code type.

File: Cargo.toml (hunk 2 above)
Comment (nrc) [reply Cargo.toml]: I think depending on the Cargo.lock file only is bad practice. It is a nice property that you can run `cargo update` and not have any breaking changes. By using bare git deps it means you have to be very careful about updating or you can get a breaking change.

> But I do think we should depend on a version once it's released, maybe 0.5.0-alpha.2.

Sure!

File: Makefile (hunk 1 above)
Comment (nrc) [reply Makefile]: :-( I guess just remove the lint for now then. It should be fixed fairly soon

File: src/import/service.rs (hunk 3 above)
Comment (nrc) [reply src/import/service.rs]: Ah, OK, shame it doesn't stick to the naming conventions I guess

File: Cargo.toml [hunk 16]
```diff
@@ -135,7 +135,7 @@ codec = { path = "components/codec" }
 tipb_helper = { path = "components/tipb_helper" }
 tipb = { git = "https://github.com/pingcap/tipb.git" }
 # TODO: use master branch after the next version is released.
```
[the remark this answers is not in the document]
Comment (BusyJay) [reply Cargo.toml]: Good catch!

File: src/server/service/kv.rs (hunk 9 above)
Comment (hunterlxt) [reply src/server/service/kv.rs]: be glad to. So wrap our i32 into such this struct ? -> [refer](https://github.com/hyperium/http/blob/master/src/status.rs#L43)

Comment (BusyJay) [reply src/server/service/kv.rs]: Yes. We should leave grpc-sys untouch and define the type in grpcio itself.

File: README.md [hunk 17]
```diff
@@ -109,7 +109,8 @@ To build TiKV you'll need to at least have the following installed:
 * `cmake` - Build tool (required for gRPC)
 * `go` - Programming language (required for gRPC)
 * `make` - Build tool (run common workflows)
-* `clang` or `gcc` - C compiler toolchain
+* `clang` or `gcc` - C++ compiler toolchain
+* `llvm` and `clang` - Used to generate bindings for different platforms (required for grpcio)
```
Comment (disksing) [inline README.md]: Can we avoid mention clang twice?

File: Cargo.toml [hunk 18]
```diff
@@ -103,8 +103,8 @@ zipf = "5.0.1"
 bitflags = "1.0.1"
 fail = "0.3"
 uuid = { version = "0.6", features = [ "serde", "v4" ] }
-grpcio = { version = "0.4", features = [ "secure" ] }
-raft = "0.4.3"
+grpcio = { version = "0.5.0-alpha.3", features = [ "secure" ] }
+raft = "0.6.0-alpha"
```
Comment (overvenus) [inline Cargo.toml]: Does it compatible with 0.4.x, in particular, the membership change?

File: src/raftstore/store/peer.rs [hunk 19]
```diff
@@ -1749,18 +1748,21 @@ impl Peer {
         match change_type {
             ConfChangeType::AddNode => {
                 if let Err(raft::Error::NotExists(_, _)) = progress.promote_learner(peer.get_id()) {
-                    let _ = progress.insert_voter(peer.get_id(), Progress::default());
+                    let _ = progress.insert_voter(peer.get_id(), Progress::new(0, 0));
                 }
             }
             ConfChangeType::RemoveNode => {
-                progress.remove(peer.get_id());
+                box_try!(progress.remove(peer.get_id()));
```
Comment (overvenus) [inline src/raftstore/store/peer.rs]: What about `?`?

File: Cargo.toml (hunk 18 above)
Comment (BusyJay) [reply Cargo.toml]: Yes, old APIs still  work.

**Comment by BusyJay** [comment]
@disksing @overvenus PTAL

File: src/coprocessor/codec/mysql/duration.rs [hunk 20]
```diff
@@ -95,7 +95,7 @@ mod parser {
 
     /// Extracts a `u32` from a buffer which matches pattern: `\d+.*`
     ///
-    /// ```compile_fail
+    /// ```ignore
```
Comment (siddontang) [inline src/coprocessor/codec/mysql/duration.rs]: why ignore here?

**Comment by siddontang** [comment]
Finally, it passed CI :-)

File: components/engine/src/lib.rs [hunk 21]
```diff
@@ -2,16 +2,7 @@
...
-    kv,
-    slog_kv,
-    slog_error,
-    slog_warn,
-    slog_record,
-    slog_b,
-    slog_log,
-    slog_record_static
-)]
+#[macro_use(slog_error, slog_warn)]
```
Comment (breeswish) [inline components/engine/src/lib.rs]: Hummm. How can these be removed without unused notice previously?

**Comment by breeswish** [comment]
Remember to open an issue to add back the removed rust 2018 idiom lint rule.

**Comment by disksing** [comment]
merge? /cc @BusyJay

**Comment by BusyJay** [comment]
Just merged master and solved the conflicts. PTAL
